log4j-jul/src/main/java/org/apache/logging/log4j/jul/Log4jBridgeHandler.java

```diff
@@ -1,319 +1,322 @@
+    public Log4jBridgeHandler() {
+        final java.util.logging.LogManager julLogMgr = java.util.logging.LogManager.getLogManager();
+        final String className = this.getClass().getName();
+        init(Boolean.parseBoolean(julLogMgr.getProperty(className + ".sysoutDebug")),
+                julLogMgr.getProperty(className + ".appendSuffix"),
+                Boolean.parseBoolean(julLogMgr.getProperty(className + ".propagateLevels")) );
+
+    }
+
+    /** Initialize this handler with given configuration. */
+    public Log4jBridgeHandler(boolean debugOutput, String suffixToAppend, boolean propagateLevels) {
+           init(debugOutput, suffixToAppend, propagateLevels);
+       }
+
+
+    /** Perform init. of this handler with given configuration (typical use is for constructor). */
+       protected void init(boolean debugOutput, String suffixToAppend, boolean propagateLevels) {
+           this.doDebugOutput = debugOutput;
+        if (debugOutput) {
+            new Exception("DIAGNOSTIC ONLY (sysout):  Log4jBridgeHandler instance created (" + this + ")")
+                    .printStackTrace(System.out);    // is no error thus no syserr
+        }
+
+        if (suffixToAppend != null) {
+            suffixToAppend = suffixToAppend.trim();    // remove spaces
+            if (suffixToAppend.isEmpty()) {
+                suffixToAppend = null;
+            } else if (suffixToAppend.charAt(0) != '.') {    // always make it a sub-logger
+                suffixToAppend = '.' + suffixToAppend;
+            }
+        }
+        this.julSuffixToAppend = suffixToAppend;
+
+        this.installAsLevelPropagator = propagateLevels;
+
+        SLOGGER.debug("Log4jBridgeHandler init. with: suffix='{}', lvlProp={}, instance={}",
+                suffixToAppend, propagateLevels, this);
+    }
+
+
+    @Override
+    public void close() {
+        // cleanup and remove listener and JUL logger references
+        julLoggerRefs = null;
+        LoggerContext.getContext(false).removePropertyChangeListener(this);
+        if (doDebugOutput) {
+            System.out.println("sysout:  Log4jBridgeHandler close(): " + this);
+        }
+    }
+
+
+    @Override
+    public void publish(LogRecord record) {
+        if (record == null) {    // silently ignore null records
+            return;
+        }
+
+        if (this.installAsLevelPropagator) {
+            synchronized(this) {
+                @SuppressWarnings("resource")    // no need to close the AutoCloseable ctx here
+                LoggerContext context = LoggerContext.getContext(false);
+                context.addPropertyChangeListener(this);
+                propagateLogLevels(context.getConfiguration());
+                // note: java.util.logging.LogManager.addPropertyChangeListener() could also
+                // be set here, but a call of JUL.readConfiguration() will be done on purpose
+                this.installAsLevelPropagator = false;
```

Comment:
By looking at some established practices for lazy initialization (e.g. [Commons Lang](https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/concurrent/LazyInitializer.java)), there should be a second `if (this.installAsLevelPropagator)` inside the synchronized block to ensure that `propagateLogLevels` is called only once.

In the `master` branch we do have a generic [LazyUtil](https://github.com/apache/logging-log4j2/blob/master/log4j-api/src/main/java/org/apache/logging/log4j/util/LazyUtil.java), but it is probably not worth backporting it for a single callsite.

Reply:
It depends if other lazy initialized code could be refactored in this branch, too. Can look for volatile fields for starters.

Reply:
`installAsLevelPropagator` must also be volatile

Reply:
Uhhhh, now it makes sense to me as well. I will modify the code tomorrow

Reply:
@ppkarwasz @carterkozak `volatile` is applied. Anything else? Thx!

Reply:
@mknet: a second check for `installAsLevelPropagator` is needed so that if two thread enter in the first conditional block one will update the levels and the other will wait until it is done:
```lang-java
if (this.installAsLevelPropagator) {
    synchronized(this) {
        if (this.installAsLevelPropagator) {
```

Reply:
Check. Will come back to you in the next couple of ours.